utilities/tools/Test-ModuleLocally.ps1

```diff
@@ -0,0 +1,136 @@
+        PrincipalId       = '12345678-1234-1234-abcd-1369d14d0d45'
+        TenantId          = '12345678-1234-1234-abcd-1369d14d0d45'
+        RemoveDeployment  = $false
+    }
+}
+
+Test-ModuleLocally @TestModuleLocallyInput
+
+.NOTES
+Make sure you provide the right information in the 'DeployParameters' parameter for this function abd Ensure you have the ability to perform the deployment operations locally.
+#>
+function Test-ModuleLocally {
+    [CmdletBinding()]
+    param (
+        [parameter(Mandatory)]
+        [string]$ModuleName,
+
+        [parameter(Mandatory = $false)]
+        [switch]$PesterTest,
+
+        [Parameter(ParameterSetName = 'Deploy')]
+        [parameter(Mandatory = $false)]
+        [switch]$DeployTest,
+
+        [Parameter(ParameterSetName = 'Deploy')]
+        [parameter(Mandatory)]
+        [psobject]$DeployParameters
+    )
+
+    begin {
+        Write-Verbose "Running Local Tests for $($ModuleName.Split('\')[-1])"
+    }
+    process {
+        # Test Module
+        if ($PesterTest) {
+            Write-Verbose '---- Performing Pester Tests ----'
+            Invoke-Pester -Configuration @{
+                Run        = @{
+                    Container = New-PesterContainer -Path (Join-Path $PSScriptRoot '..\..' 'arm\.global\global.module.tests.ps1') -Data @{
+                        moduleFolderPaths = (Join-Path $PSScriptRoot '..\..\arm' $ModuleName)
+                    }
+                }
+                Filter     = @{
+                    #ExcludeTag = 'ApiCheck'
+                    #Tag = 'ApiCheck'
+                }
+                TestResult = @{
+                    TestSuiteName = 'Global Module Tests'
+                    Enabled       = $false
+                }
+                Output     = @{
+                    Verbosity = 'Detailed'
+                }
+            }
+        }
+        # Deploy Module
+        if ($DeployTest -and $DeployParameters) {
+            Write-Verbose '---- Deploying Module ----'
+            # Find Test Parameter Files
+            $ModuleParameterFiles = Get-ChildItem -Path (Join-Path $PSScriptRoot '..\..\arm' $ModuleName '.parameters') -Recurse
+            # Load Tokens Converter Script
+            . (Join-Path $PSScriptRoot '..\..' '.github\actions\sharedScripts\Convert-TokensInFiles.ps1')
+            # Replace Tokens with Values For Local Testing
+            $TokensReplaceWithObject = @(
+                @{ Replace = '<<tenantId>>'; With = "$($DeployParameters.TenantId)" }
+                @{ Replace = '<<subscriptionId>>'; With = "$($DeployParameters.SubscriptionId)" }
+                @{ Replace = '<<managementGroupId>>'; With = "$($DeployParameters.ManagementGroupId)" }
+                @{ Replace = '<<principalId>>'; With = "$($DeployParameters.PrincipalId)" }
+            )
+            $ModuleParameterFiles | ForEach-Object { Convert-TokensInFiles -Paths $PSitem.FullName -TokensReplaceWith $TokensReplaceWithObject }
```

Comment:
I should have caught that last time already but PowerShell functions should use singular names (i.e. `Convert-TokensInFile`) or if you want to imply the plural nature `Convert-TokensInFileList`

Reply:
That's ok. I can rename that, already going to modify the Deploy and Validate actions for new tokens anyways. Thanks for calling it out.

Reply:
Resolved. and added support for custom output path